migrations/20231128_jellyfish_migration/utils/data_migration.go

```diff
@@ -0,0 +1,386 @@
+package utils
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+)
+
+type UpdateData struct {
+	Filter    interface{} `json:"-"`
+	ItemID    string      `json:"_id"`
+	UserID    string      `json:"_userId"`
+	ItemType  string      `json:"-"`
+	Apply     []bson.M    `json:"apply"`
+	ApplyLast bson.M      `json:"applyLast"`
+	Revert    []bson.M    `json:"revert"`
+}
+
+type ErrorData struct {
+	Error    error  `json:"error"`
+	ItemID   string `json:"_id"`
+	ItemType string `json:"-"`
+	Msg      string `json:"message,omitempty"`
+}
+
+type MigrationStats struct {
+	Errored int
+	Fetched int
+	Applied int
+	ToApply int
+	Elapsed time.Duration
+}
+
+type Settings struct {
+	DryRun              bool
+	Rollback            bool
+	RollbackSectionName string
+	StopOnErr           bool
+	WriteBatchSize      *int64
+	QueryBatchSize      int
+	QueryBatchLimit     int
+
+	capacity    *int
+	writeToDisk bool
+}
+
+func NewSettings(dryRun *bool, stopOnErr *bool, rollback *bool, rollbackSectionName *string, capacity *int, queryBatch *int, queryLimit *int, writeToDisk *bool) *Settings {
+	settings := &Settings{
+		writeToDisk:         false,
+		Rollback:            true,
+		RollbackSectionName: "_rollbackMigration",
+		DryRun:              true,
+		StopOnErr:           true,
+		QueryBatchSize:      50,
+		QueryBatchLimit:     100,
+	}
+	if dryRun != nil {
+		settings.DryRun = *dryRun
+	}
+	if stopOnErr != nil {
+		settings.StopOnErr = *stopOnErr
+	}
+	if rollback != nil {
+		settings.Rollback = *rollback
+	}
+	if rollbackSectionName != nil {
+		settings.RollbackSectionName = *rollbackSectionName
+	}
+	if writeToDisk != nil {
+		settings.writeToDisk = *writeToDisk
+	}
+	if queryBatch != nil {
+		settings.QueryBatchSize = *queryBatch
+	}
+	if queryLimit != nil {
+		settings.QueryBatchLimit = *queryLimit
+	}
+	if capacity != nil && *capacity > 0 {
+		settings.capacity = capacity
+		log.Printf("capped at %d items", *settings.capacity)
+	}
+	return settings
+}
+
+type DataMigration struct {
+	ctx                  context.Context
+	dataC                *mongo.Collection
+	settings             *Settings
+	updates              []mongo.WriteModel
+	groupedDiffs         map[string][]UpdateData
+	groupedErrors        groupedErrors
+	errorsCount          int
+	updatedCount         int
+	fetchedCount         int
+	lastUpdatedID        string
```

Comment:
Suggest you just make this a `*string` to avoid any of the `pointer.String` stuff.

Reply:
Besides, when not using a pointer, the default is an empty string, which you have a special case to check for in `jellyfishQuery`. Just use a pointer and get rid of an empty string logic.